src/rpc/blockchain.cpp

```diff
@@ -1501,14 +1502,79 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
     ret.pushKV("unbroadcastcount", uint64_t{pool.GetUnbroadcastTxs().size()});
+
+    if (feeLimits) {
+        const MempoolHistogramFeeLimits& limits = feeLimits.value();
+
+        /* Keep histogram per...
+         * ... cumulated tx sizes
+         * ... txns (count)
+         * ... cumulated fees */
+        std::vector<uint64_t> sizes(limits.size(), 0);
+        std::vector<uint64_t> count(limits.size(), 0);
+        std::vector<uint64_t> fees(limits.size(), 0);
+
+        for (const CTxMemPoolEntry& e : pool.mapTx) {
+            int size = (int)e.GetTxSize();
+            CAmount fee = e.GetFee();
+            uint64_t asize = e.GetSizeWithAncestors();
+            CAmount afees = e.GetModFeesWithAncestors();
+            uint64_t dsize = e.GetSizeWithDescendants();
+            CAmount dfees = e.GetModFeesWithDescendants();
+
+            CAmount fpb = fee / size; // fee per byte
+            CAmount afpb = afees / asize; // fee per byte including ancestors
+            CAmount dfpb = dfees / dsize; // fee per byte including descendants
```

Comment:
@jonatack This is a part of the original PR I need to research more.

Do you possibly have a tip where to learn how to properly compute `feeperbyte` value for a transaction? Or possibly who would know that? 

Anyway, I'm slowly skimming Bitcoin Core codebase so maybe I'll be lucky :-)

Reply:
For computing a fee rate per byte, have a look in `src/policy/feerate.{h, cpp}`. Here's a proposed diff (using the current code) to do explicit casts rather than implicit conversions and narrowing. The size denominators are calling `GetTxSize()`, which should be non-zero, but you can `CHECK_NONFATAL` on these. That said, you may be right that it can be done properly/better, as you asked, by using the `CFeeRate` ctors.

<details><summary>diff</summary><p>

```diff
         for (const CTxMemPoolEntry& e : pool.mapTx) {
-            const int size = (int)e.GetTxSize();
-            CAmount fee = e.GetFee();
-            uint64_t asize = e.GetSizeWithAncestors();
-            CAmount afees = e.GetModFeesWithAncestors();
-            uint64_t dsize = e.GetSizeWithDescendants();
-            CAmount dfees = e.GetModFeesWithDescendants();
-
-            CAmount fpb = fee / size; // fee per byte
-            CAmount afpb = afees / asize; // fee per byte including ancestors
-            CAmount dfpb = dfees / dsize; // fee per byte including descendants
-            CAmount tfpb = (afees + dfees - fee) / (asize + dsize - size);
-            CAmount feeperbyte = std::max(std::min(dfpb, tfpb), std::min(fpb, afpb));
+            const CAmount fee{e.GetFee()};
+            const CAmount afees{e.GetModFeesWithAncestors()};
+            const CAmount dfees{e.GetModFeesWithDescendants()};
+
+            const int64_t size{static_cast<int64_t>(e.GetTxSize())};
+            const int64_t asize{static_cast<int64_t>(e.GetSizeWithAncestors())};
+            const int64_t dsize{static_cast<int64_t>(e.GetSizeWithDescendants())};
+
+            CHECK_NONFATAL(size > 0);
+            CHECK_NONFATAL(asize > 0);
+            CHECK_NONFATAL(dsize > 0);
+            CHECK_NONFATAL(asize + dsize - size > 0);
+
+            CAmount fpb{fee / size};     // fee per byte
+            CAmount afpb{afees / asize}; // fee per byte including ancestors
+            CAmount dfpb{dfees / dsize}; // fee per byte including descendants
+            CAmount tfpb{(afees + dfees - fee) / (asize + dsize - size)};
+            CAmount feeperbyte{std::max(std::min(dfpb, tfpb), std::min(fpb, afpb))};
```
</p></details>

Reply:
(note, `CAmount` is `int64_t`, see `src/amount.h`)

Reply:
Thanks. Will have a look!

Reply:
Ah, my comment didn't make it through the internet:

It would be good to use feerate values in sat/vB for both the user input and the output here, as we are slowly moving from BTC/kvB to sat/vB fee rate units, per user demand.

Currently, you can construct a feerate from a sat/vB amount with `CFeeRate{amount, COIN}`. I proposed `CFeeRate::FromSatB` and `FromBtcKb` named constructors in #20790, #20391, #20546 but none of the three are/were merged.

Reply:
Originally, fee rates were in sats but then I found https://github.com/bitcoin/bitcoin/pull/12240 which made me think that sats -> BTC is preferred.

Reply:
So I will switch it back to use sats. Thanks for the `CFeeRate{amount, COIN}` tip!

Reply:
@jonatack Do I understand correctly that `CAmount fpb = fee / size` is equivalent with `CAmount fpb = CFeeRate(fee, size).GetFee(1)`? Is that what you propose to use?

Reply:
You can verify this with, for example
```cpp
            assert(fpb == CFeeRate(fee, size).GetFee(1));
            assert(afpb == CFeeRate(afees, asize).GetFee(1));
            assert(dfpb == CFeeRate(dfees, dsize).GetFee(1));
            assert(tfpb == CFeeRate(afees + dfees - fee, asize + dsize - size).GetFee(1));
```
(I left this as a hint in the suggestions in https://github.com/bitcoin/bitcoin/pull/21422#pullrequestreview-621285483, but please remove it and decide what you think is best)

Reply:
Thanks! I like `CFeeRate` better as then I don't add new divisions myself. The asserts work for me okay.